### DISCUSSION TITLE
Error running Example 1a of tutorial "Unable to connect to broker->unable to register federate"

### DISCUSSION THREAD

**Discussion by adelheidari** [GMLC-TDC/HELICS#2571] [General]
Hi everyone,
I'm trying to run the first example in tutorial (Example 1a) which co-simulates a transmission system (in Pypower) and distribution system (in GridLab-D). This is the link for the example files: https://github.com/GMLC-TDC/HELICS-Examples/tree/main/user_guide_examples/misc/gridlabd_example_1/Example_1a
I'm getting the following error in TransmissionFederate log file: 

`[2023-11-03 10:57:57.473] [console] [warning] commWarning||Transmission Federate (0)[connecting]::zmq broker connection timed out, trying again (2)
[2023-11-03 10:57:57.474] [console] [warning] commWarning||Transmission Federate (0)[connecting]::sending message to tcp://127.0.0.1:23405
[2023-11-03 10:58:17.503] [console] [error] commERROR||Transmission Federate (0)[connecting]::zmq broker connection timed out after trying 5 times (2)
[2023-11-03 10:58:17.503] [console] [error] commERROR||Transmission Federate (0)[connecting]::receiver connection failure
Traceback (most recent call last):
  File "C:\Users\Ssus\Python\HELICS\Example_1a\Transmission\Transmission_simulator.py", line 53, in <module>
    fed = h.helicsCreateValueFederateFromConfig("Transmission_json.json")
  File "C:\Users\Ssus\AppData\Local\Programs\Python\Python310\lib\site-packages\helics\capi.py", line 3635, in helicsCreateValueFederateFromConfig
    raise HelicsException("[" + str(err.error_code) + "] " + ffi.string(err.message).decode())
helics.capi.HelicsException: [-1] Unable to connect to broker->unable to register federate`

And the Distribution federate log file is empty. It is stopping at the very beginning line of the code in Transmission_simulator.py (line 53) which trying to create a federate. Here are the versions of the software I'm using: 
python v3.9.16
helics v 3.4.0
GridLab-D v 5.2
Pypower v5.1.16

I would really appreciate if you could help me to fix this issue.

**Comment by trevorhardy** [comment]
How are you launching the co-simulation?

**Reply by adelheidari** [reply]
Thanks for the reply. 

I’m running the co-simulation through command prompt. I’m running “cosim_runner_example_1a” using helics run —path cosim_runner_example_1a.json in the command prompt.

**Reply by trevorhardy** [reply]
It looks like we have two ways of launching this that are not compatible. In the cosim_runner_example_1a.json, set "broker" to "true" and try running that. I suspect that will work and if it does, I can make the change in the repository.

**Reply by adelheidari** [reply]
I ran the case with the edit you mentioned but it still shows the same error.

**Reply by trevorhardy** [reply]
Have you tried running the Python-based examples ([such as this one](https://github.com/GMLC-TDC/HELICS-Examples/tree/main/user_guide_examples/fundamental/fundamental_default))? Do you get the same-error?

**Reply by adelheidari** [reply]
I ran the example you mentioned, and it is running without any error. So, what is wrong with example 1a? Could you please explain so that I can fix the problem with example 1a?

**Reply by trevorhardy** [reply]
Ok, I should have started with this; I suspect you have a broker already running from a previous run. Look for and kill all helics_brokers that are currently running before trying to re-run the example. There should be no helics_brokers running when you start to try to launch the co-simulation.

**Reply by adelheidari** [reply]
I run helics list-brokers and it shows nothing in the output. Then, I run helics kill-all-brokers and it says: ERROR: The process "helics_broker.exe" not found.

**Reply by adelheidari** [reply]
This is the message showing up in cmd when I'm running example 1a. I wonder it maybe helpful. 
`C:\Users\Ssus\Python\HELICS\Example_1a>helics run --path cosim_runner_Example_1a.json

[info] Running federation: Example-1a-T-D-Cosimulation-HELICSRunner
[info] Running federate TransmissionFederate as a background process
[info] Running federate GridLABDFederate as a background process
Traceback (most recent call last):
  File "C:\Users\Ssus\AppData\Local\Programs\Python\Python310\lib\runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:\Users\Ssus\AppData\Local\Programs\Python\Python310\lib\runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "C:\Users\Ssus\AppData\Local\Programs\Python\Python310\Scripts\helics.exe\__main__.py", line 7, in <module>
  File "C:\Users\Ssus\AppData\Local\Programs\Python\Python310\lib\site-packages\click\core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
  File "C:\Users\Ssus\AppData\Local\Programs\Python\Python310\lib\site-packages\click\core.py", line 1078, in main
    rv = self.invoke(ctx)
  File "C:\Users\Ssus\AppData\Local\Programs\Python\Python310\lib\site-packages\click\core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "C:\Users\Ssus\AppData\Local\Programs\Python\Python310\lib\site-packages\click\core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "C:\Users\Ssus\AppData\Local\Programs\Python\Python310\lib\site-packages\click\core.py", line 783, in invoke
    return __callback(*args, **kwargs)
  File "C:\Users\Ssus\AppData\Local\Programs\Python\Python310\lib\site-packages\helics\cli.py", line 260, in run
    p = subprocess.Popen(
  File "C:\Users\Ssus\AppData\Local\Programs\Python\Python310\lib\subprocess.py", line 971, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "C:\Users\Ssus\AppData\Local\Programs\Python\Python310\lib\subprocess.py", line 1456, in _execute_child
    hp, ht, pid, tid = _winapi.CreateProcess(executable, args,
OSError: [WinError 193] %1 is not a valid Win32 application`

**Reply by trevorhardy** [reply]
The runner may be having a problem for some reason. Have you tried running the three federates individually in three separate command windows?

`helics_broker -f2`
(In the Distribution folder) `gridlabd IEEE_123_feeder_0.glm`
(In the Transmission folder) `python Transmission_simulator.py`

**Reply by adelheidari** [reply]
I run the codes you mentioned. I am new in this field and using helics, but it seems that Transmission federate is running, but there is a problem with gridlabd. Correct me if I'm wrong. 

Outputs:
`C:\Users\Ssus\Python\HELICS\Example_1a\Transmission>python Transmission_simulator.py
HELICS Version: 3.4.0 (2023-01-19)
TransmissionSim: Federate TransmissionSim has been registered
TransmissionSim: Registered Publication ---> TransmissionSim/transmission_voltage
C:\Users\Ssus\AppData\Roaming\Python\Python39\site-packages\helics\capi.py:9014: UserWarning: This is deprecated. Use `helicsInputGetTarget` instead.
  warnings.warn("This is deprecated. Use `helicsInputGetTarget` instead.")
TransmissionSim: Registered Subscription ---> IEEE_123_feeder_0/totalLoad`

`C:\Users\Ssus\Python\HELICS\Example_1a\Distribution>gridlabd IEEE_123_feeder_0.glm

ERROR    [INIT] : D:/a/gridlab-d/gridlab-d/gridlab-build/gldcore/module.cpp(547): module 'connection' load failed - "C:\Program Files\GridLAB-D 5.2.0\bin\connection.dll": The specified module could not be found.
IEEE_123_feeder_0.glm(23): module 'connection' load failed, No such file or directory
IEEE_123_feeder_0.glm(23): load failed at or near 'module conne...'
ERROR    [INIT] : unable to load 'IEEE_123_feeder_0.glm': No such file or directory
FATAL    [INIT] : shutdown after command line rejected`

**Reply by trevorhardy** [reply]
Yeah, something is wrong with the GridLAB-D installation. Does connection.dll exist in that location? Have you set GLPATH?

**Reply by trevorhardy** [reply]
@afisher1, can you offer any help on what be going on with this GridLAB-D install.

**Reply by adelheidari** [reply]
Yes, connection.dll exists in the bin folder. Could you explain what do you mean by GLPATH?

**Comment by afisher1** [comment]
Hi all,
Typically when that happens it's because GridLAB-D couldn't find the libhelics.dll/so/dylib. It looks like you are on Windows so I recommend checking to see if your user/system level PATH environment variable contains the path to "HELICS install location"\bin and/or \lib.

**Reply by adelheidari** [reply]
Hi,

Thanks for the reply. I have added the GridLab-D\bin path and helics package installed location in python (both install/lib and install/bin folders) to the local and system Path variable. But, it still shows the same error.

**Comment by afisher1** [comment]
So in the cmd terminal where you run gridlabd can you confirm that the path to libhelics.dll existing the %PATH% environment variable? You shouldn't need to worry about creating a GLPATH environment variable if you are using v5.2 of gridlabd.

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Reply by adelheidari** [reply]
There is not any file named "libhelics.dll" in the helics package. I thought you mean lib\helics.dll in the previous message. There are "helics.dll" and "helics.lib" in helics\install\bin and lib folders.

**Reply by adelheidari** [reply]
To answer your question, I did this: I ran 'where helics.dll' and 'where helics.lib' in command window and the output paths of these commands are in the %PATH% (I ran 'echo %PATH%' in cmd to see what is included in %PATH%).

**Reply by adelheidari** [reply]
@afisher1, Should I enable HELICS in GridLAB-D manually? I saw that HELICS is enabled after version 4.3 of GridLAB-D and there is no need to do any other stuff in this regard.

**Comment by afisher1** [comment]
I apologize for the delayed response. Can you try renaming helics.dll to libhelics.dll? No need to rename helics.lib. I get your behavior when the library is helics.dll and I got it to work by changing the filename.

**Reply by adelheidari** [reply]
Thank you so much for the help. I really appreciate it. 
I was able to run the case with your suggestion. Also, there were a few other minor corrections needed to be done in order to run the case. I found out that new files have been uploaded in the Github ([here](https://github.com/GMLC-TDC/HELICS-Examples/tree/main/user_guide_examples/misc/gridlabd_example_1)) for the examples. Using those files along with the suggestion you mentioned, the case can be successfully run. 

Thanks again,
Adel

**Reply by trevorhardy** [reply]
Great; I'll be closing this out, then. Thanks for trying that example out and helping us get it back up to snuff.
